keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1045 @@
+* A new TLS user-mapping capability in kube-apiserver to understand client
+  certificates that embed PodIdentity extensions.
+* A new approver/signer pair, shipped in kube-controller-manager, to issue
+  compatible client certificate to pods.
+* An enhancement to client-go's InClusterConfig that automatically loads a
+  credential bundle from a well-known filesystem path and uses is as a client
+  certificate.
+
+There is a draft implementation of the design outlined in this KEP:
+https://github.com/kubernetes/kubernetes/pull/121596
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+**Ensure that certificate issuance is accomplished without relying on bearer tokens at any step**:  One of the most significant benefits of mTLS authentication is that the credential is split into two pieces (private key and certificate), only one of which is ever transmitted to the counterparty (and thus subject to eavesdropping or theft by the counterparty).  In order to preserve this guarantee, it's important that bearer tokens are not involved *at any step*, otherwise the security of the scheme devolves to just being a bearer token.  In this KEP, as long as both kubelet and the signer implementation authenticate to kube-apiserver using their own separately-bootstrapped client certificates, then no bearer tokens are involved in certificate issuance.
+
+**Make it easy to ship secure third-party signer implementations**:  Securely implementing and operating a third-party signer implementation today requires intimate familiarity with Kubernetes' security model, otherwise your signer may negate Kubernetes' node restriction security boundary.  This KEP moves responsibilty for enforcing node restriction on CSRs into kube-apiserver, so that signer implementations do not need to consider it.
+
+### Non-Goals
+
+**Specify a solution for pod-to-pod mTLS in core Kubernetes**: Pod-to-pod mTLS (aka service mesh) is a rapidly-evolving area.  While the PodCertificate volume and PodIdentity X.509 extension are expected (and designed to be) useful for service mesh implementations, it doesn't make sense to try and standardize pod-to-pod mTLS at the core Kubernetes layer today.
+
+**Support TPM-backed private keys**: In this proposal, private keys are
+contained to the node (held only in kubelet memory and tmpfs projected volumes),
+but they can still be extracted from a compromised workload (or compromised
+node).  Storing the private keys in TPMs closes this loophole, but will require
```

Comment:
I wasn't being precise here, sorry.  My intention was just to vaguely handwave at some sort of HSM support.

However, I have been thinking a little bit more about this as I'm preparing to ramp back up on this after working on external token signing.  I am starting to think that it might be a good idea to *not* write the key into the pod filesystem, and instead expose some external signature interface via UDS in the pod.   Basically, start by requiring external TLS.

To establish a TLS connection to kube-apiserver, we would ship a custom crypto.Signer implementation with client-go that knows how to speak to kubelet via the UDS and request signatures of data using the per-pod key held by kubelet.

In Alpha, kubelet wouldn't have to *do* anything fancy with the keys, but starting out with the key not exposed to the pod will make an eventual TPM / HSM transition transparent, and it will also significantly improve the baseline security of the feature.

Reply:
After API review, we decided to defer any HSM / external signing  support to a future iteration.